[Pytorch][Annotation] Update inlined callstack with module instance info

cccclai opened the pull request:
### Summary
This PR refers to this pr: https://github.com/pytorch/pytorch/pull/42174
Copy the motivation from the pr:
> Motivation: During inlinng we lose information about where the node/op is coming from in module hierarchy. The SourceDebug info is useful in providing where in the source code the definition of the node exist. One can follow the callstack to find the relevant module hierarchy. However, this is less accessible, particularly in the context of aggregating node level ops latencies to produce module level latencies. Facilitating this enables identifying which module might be expensive and can be targeted either for optimization or redesign. This also aids network architecture search approaches.

As the first step, this PR adds module instance information to inlined call stack during inlining. Next step, find the module path by traversing callstack.

### Test Plan
CI

/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=LiteInterpreterTest.DuplicatedClassTypeModuleInfo:LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo:*/LiteInterpreterTest.DuplicatedClassTypeModuleInfo/*:*/LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo --gtest_color=no

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46729 [Pytorch][Annotation] Update inlined callstack with module instance info**

Differential Revision: [D24493220](https://our.internmc.facebook.com/intern/diff/D24493220)

cccclai committed:
[Pytorch][Annotation] Update inlined callstack with module instance info

[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Update inlined callstack with module instance info"




[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Update inlined callstack with module instance info"


Differential Revision: [D24493220](https://our.internmc.facebook.com/intern/diff/D24493220)

[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Update inlined callstack with module instance info"

### Summary
This PR refers to this pr: https://github.com/pytorch/pytorch/pull/42174
Copy the motivation from the pr:
> Motivation: During inlinng we lose information about where the node/op is coming from in module hierarchy. The SourceDebug info is useful in providing where in the source code the definition of the node exist. One can follow the callstack to find the relevant module hierarchy. However, this is less accessible, particularly in the context of aggregating node level ops latencies to produce module level latencies. Facilitating this enables identifying which module might be expensive and can be targeted either for optimization or redesign. This also aids network architecture search approaches.

As the first step, this PR adds module instance information to inlined call stack during inlining. Next step, find the module path by traversing callstack.

### Test Plan
/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=LiteInterpreterTest.DuplicatedClassTypeModuleInfo:LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo:*/LiteInterpreterTest.DuplicatedClassTypeModuleInfo/*:*/LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo --gtest_color=no


Differential Revision: [D24493220](https://our.internmc.facebook.com/intern/diff/D24493220)

[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Update inlined callstack with module instance info"

### Summary
This PR refers to this pr: https://github.com/pytorch/pytorch/pull/42174
Copy the motivation from the pr:
> Motivation: During inlinng we lose information about where the node/op is coming from in module hierarchy. The SourceDebug info is useful in providing where in the source code the definition of the node exist. One can follow the callstack to find the relevant module hierarchy. However, this is less accessible, particularly in the context of aggregating node level ops latencies to produce module level latencies. Facilitating this enables identifying which module might be expensive and can be targeted either for optimization or redesign. This also aids network architecture search approaches.

As the first step, this PR adds module instance information to inlined call stack during inlining. Next step, find the module path by traversing callstack.

### Test Plan
/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=LiteInterpreterTest.DuplicatedClassTypeModuleInfo:LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo:*/LiteInterpreterTest.DuplicatedClassTypeModuleInfo/*:*/LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo --gtest_color=no


Differential Revision: [D24493220](https://our.internmc.facebook.com/intern/diff/D24493220)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/ir/scope.h:
```diff
@@ -50,6 +51,26 @@ struct TORCH_API Scope : public c10::intrusive_ptr_target {
+  c10::ClassTypePtr class_type() {
+    return module_type_;
+  }
+  std::string instance_name() {
```
Nit: What do we need a non-const version of this method for?

kimishpatel replied:
Yeah this may not be needed. If so please remove.

ZolotukhinM commented on torch/csrc/jit/ir/scope.h:
```diff
@@ -1,5 +1,6 @@
 #pragma once
 #include <ATen/core/interned_strings.h>
+#include <ATen/core/jit_type.h>
```
Nit: A forward declaration of `c10::ClassTypePtr` could probably suffice in this case I think and would allow to avoid including a heavy header file.

ZolotukhinM commented on torch/csrc/jit/ir/scope.h:
```diff
@@ -50,6 +51,26 @@ struct TORCH_API Scope : public c10::intrusive_ptr_target {
 
 struct Function;
 struct InlinedCallStack;
+struct ModuleInstanceInfo {
```
Nit: It'd be good to have some comments for the class and its public methods.

cccclai replied:
Address the comment in latest commit (d6403b92e3ceb7708c3264e1b9875163d2a2658e).

ZolotukhinM reviewed (approved):
Overall looks good to me!

cccclai committed:
Update on "[Pytorch][Annotation] Update inlined callstack with module instance info"

### Summary
This PR refers to this pr: https://github.com/pytorch/pytorch/pull/42174
Copy the motivation from the pr:
> Motivation: During inlinng we lose information about where the node/op is coming from in module hierarchy. The SourceDebug info is useful in providing where in the source code the definition of the node exist. One can follow the callstack to find the relevant module hierarchy. However, this is less accessible, particularly in the context of aggregating node level ops latencies to produce module level latencies. Facilitating this enables identifying which module might be expensive and can be targeted either for optimization or redesign. This also aids network architecture search approaches.

As the first step, this PR adds module instance information to inlined call stack during inlining. Next step, find the module path by traversing callstack.

### Test Plan
CI

/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=LiteInterpreterTest.DuplicatedClassTypeModuleInfo:LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo:*/LiteInterpreterTest.DuplicatedClassTypeModuleInfo/*:*/LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo --gtest_color=no


Differential Revision: [D24493220](https://our.internmc.facebook.com/intern/diff/D24493220)

[ghstack-poisoned]

dhruvbird commented on torch/csrc/jit/ir/scope.h:
```diff
@@ -50,6 +51,26 @@ struct TORCH_API Scope : public c10::intrusive_ptr_target {
+  c10::ClassTypePtr class_type() const {
+    return module_type_;
+  }
+  std::string instance_name() const {
```
nit: Possibly consider returning `std::string const&` instead to avoid a copy.

kimishpatel replied:
Generally agree, but am vary of returning reference to class variable since client code may just do `auto name = ..` and use it or pass as a reference further around. In the meantime original class maybe destroyed or mutated? Or client code itself may mutate class variable if return type is non-const.

cccclai replied:
Looks like a dangling reference could be dangerous. Will use return by value here.

dhruvbird commented on torch/csrc/jit/ir/scope.h:
```diff
@@ -50,6 +51,26 @@ struct TORCH_API Scope : public c10::intrusive_ptr_target {
 struct InlinedCallStack;
+struct ModuleInstanceInfo {
+ private:
+  c10::ClassTypePtr module_type_{nullptr};
```
I'm not sure if you need to initialize it to `{nullptr}` if you are going to be assigning this in the member constructor list of the constructor.

cccclai replied:
After a second thought, `c10::ClassTypePtr` doesn't have a default constructor and it might be error-prone when it's not used correctly. Having a default nullptr here is probably a safer option.

kimishpatel commented:
Looks fine to me. Make sure you address Mikhail's comments.

kimishpatel commented:
Also address all the lint and build failures.

iseeyuan reviewed (commented):
> Also address all the lint and build failures.

@cccclai please run `python tools/clang_format_all.py` and `python tools/clang_tidy.py` to address the clang issues.

kimishpatel commented on torch/csrc/jit/ir/ir.cpp:
```diff
@@ -1897,6 +1897,19 @@ std::vector<Value*> inlineCallTo(
       new_callstack_entries;
 
+  c10::optional<ModuleInstanceInfo> module_instance_info = c10::nullopt;
+  if (to_replace->kind() == prim::CallMethod) {
```
I think CallFunction nodes are like free functions. CallMethods are methods on some objects and we get the objects via GetAttr nodes. If it is a CallFunction it can be any function. That function maybe called from some method (CallMethod) of some object but that information we will have captured when we were inlining the node corresponding to that method.
See: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/interpreter.cpp#L930-L940

cccclai committed:
Update on "[Pytorch][Annotation] Update inlined callstack with module instance info"

### Summary
This PR refers to this pr: https://github.com/pytorch/pytorch/pull/42174
Copy the motivation from the pr:
> Motivation: During inlinng we lose information about where the node/op is coming from in module hierarchy. The SourceDebug info is useful in providing where in the source code the definition of the node exist. One can follow the callstack to find the relevant module hierarchy. However, this is less accessible, particularly in the context of aggregating node level ops latencies to produce module level latencies. Facilitating this enables identifying which module might be expensive and can be targeted either for optimization or redesign. This also aids network architecture search approaches.

As the first step, this PR adds module instance information to inlined call stack during inlining. Next step, find the module path by traversing callstack.

### Test Plan
CI

/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=LiteInterpreterTest.DuplicatedClassTypeModuleInfo:LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo:*/LiteInterpreterTest.DuplicatedClassTypeModuleInfo/*:*/LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo --gtest_color=no


Differential Revision: [D24493220](https://our.internmc.facebook.com/intern/diff/D24493220)

[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Update inlined callstack with module instance info"

### Summary
This PR refers to this pr: https://github.com/pytorch/pytorch/pull/42174
Copy the motivation from the pr:
> Motivation: During inlinng we lose information about where the node/op is coming from in module hierarchy. The SourceDebug info is useful in providing where in the source code the definition of the node exist. One can follow the callstack to find the relevant module hierarchy. However, this is less accessible, particularly in the context of aggregating node level ops latencies to produce module level latencies. Facilitating this enables identifying which module might be expensive and can be targeted either for optimization or redesign. This also aids network architecture search approaches.

As the first step, this PR adds module instance information to inlined call stack during inlining. Next step, find the module path by traversing callstack.

### Test Plan
CI

/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=LiteInterpreterTest.DuplicatedClassTypeModuleInfo:LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo:*/LiteInterpreterTest.DuplicatedClassTypeModuleInfo/*:*/LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo --gtest_color=no


Differential Revision: [D24493220](https://our.internmc.facebook.com/intern/diff/D24493220)

[ghstack-poisoned]

cccclai reviewed (commented):
Address comments.

cccclai committed:
Update on "[Pytorch][Annotation] Update inlined callstack with module instance info"

### Summary
This PR refers to this pr: https://github.com/pytorch/pytorch/pull/42174
Copy the motivation from the pr:
> Motivation: During inlinng we lose information about where the node/op is coming from in module hierarchy. The SourceDebug info is useful in providing where in the source code the definition of the node exist. One can follow the callstack to find the relevant module hierarchy. However, this is less accessible, particularly in the context of aggregating node level ops latencies to produce module level latencies. Facilitating this enables identifying which module might be expensive and can be targeted either for optimization or redesign. This also aids network architecture search approaches.

As the first step, this PR adds module instance information to inlined call stack during inlining. Next step, find the module path by traversing callstack.

### Test Plan
CI

/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=LiteInterpreterTest.DuplicatedClassTypeModuleInfo:LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo:*/LiteInterpreterTest.DuplicatedClassTypeModuleInfo/*:*/LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo --gtest_color=no


Differential Revision: [D24493220](https://our.internmc.facebook.com/intern/diff/D24493220)

[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Update inlined callstack with module instance info"

### Summary
This PR refers to this pr: https://github.com/pytorch/pytorch/pull/42174
Copy the motivation from the pr:
> Motivation: During inlinng we lose information about where the node/op is coming from in module hierarchy. The SourceDebug info is useful in providing where in the source code the definition of the node exist. One can follow the callstack to find the relevant module hierarchy. However, this is less accessible, particularly in the context of aggregating node level ops latencies to produce module level latencies. Facilitating this enables identifying which module might be expensive and can be targeted either for optimization or redesign. This also aids network architecture search approaches.

As the first step, this PR adds module instance information to inlined call stack during inlining. Next step, find the module path by traversing callstack.

### Test Plan
CI

/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=LiteInterpreterTest.DuplicatedClassTypeModuleInfo:LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo:*/LiteInterpreterTest.DuplicatedClassTypeModuleInfo/*:*/LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo --gtest_color=no


Differential Revision: [D24493220](https://our.internmc.facebook.com/intern/diff/D24493220)

[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Update inlined callstack with module instance info"

### Summary
This PR refers to this pr: https://github.com/pytorch/pytorch/pull/42174
Copy the motivation from the pr:
> Motivation: During inlinng we lose information about where the node/op is coming from in module hierarchy. The SourceDebug info is useful in providing where in the source code the definition of the node exist. One can follow the callstack to find the relevant module hierarchy. However, this is less accessible, particularly in the context of aggregating node level ops latencies to produce module level latencies. Facilitating this enables identifying which module might be expensive and can be targeted either for optimization or redesign. This also aids network architecture search approaches.

As the first step, this PR adds module instance information to inlined call stack during inlining. Next step, find the module path by traversing callstack.

### Test Plan
CI

/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=LiteInterpreterTest.DuplicatedClassTypeModuleInfo:LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo:*/LiteInterpreterTest.DuplicatedClassTypeModuleInfo/*:*/LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo --gtest_color=no


Differential Revision: [D24493220](https://our.internmc.facebook.com/intern/diff/D24493220)

[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Update inlined callstack with module instance info"

### Summary
This PR refers to this pr: https://github.com/pytorch/pytorch/pull/42174
Copy the motivation from the pr:
> Motivation: During inlinng we lose information about where the node/op is coming from in module hierarchy. The SourceDebug info is useful in providing where in the source code the definition of the node exist. One can follow the callstack to find the relevant module hierarchy. However, this is less accessible, particularly in the context of aggregating node level ops latencies to produce module level latencies. Facilitating this enables identifying which module might be expensive and can be targeted either for optimization or redesign. This also aids network architecture search approaches.

As the first step, this PR adds module instance information to inlined call stack during inlining. Next step, find the module path by traversing callstack.

### Test Plan
CI

/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=LiteInterpreterTest.DuplicatedClassTypeModuleInfo:LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo:*/LiteInterpreterTest.DuplicatedClassTypeModuleInfo/*:*/LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo --gtest_color=no


Differential Revision: [D24493220](https://our.internmc.facebook.com/intern/diff/D24493220)

[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Update inlined callstack with module instance info"

### Summary
This PR refers to this pr: https://github.com/pytorch/pytorch/pull/42174
Copy the motivation from the pr:
> Motivation: During inlinng we lose information about where the node/op is coming from in module hierarchy. The SourceDebug info is useful in providing where in the source code the definition of the node exist. One can follow the callstack to find the relevant module hierarchy. However, this is less accessible, particularly in the context of aggregating node level ops latencies to produce module level latencies. Facilitating this enables identifying which module might be expensive and can be targeted either for optimization or redesign. This also aids network architecture search approaches.

As the first step, this PR adds module instance information to inlined call stack during inlining. Next step, find the module path by traversing callstack.

### Test Plan
CI

/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=LiteInterpreterTest.DuplicatedClassTypeModuleInfo:LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo:*/LiteInterpreterTest.DuplicatedClassTypeModuleInfo/*:*/LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo --gtest_color=no


Differential Revision: [D24493220](https://our.internmc.facebook.com/intern/diff/D24493220)

[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Update inlined callstack with module instance info"

### Summary
This PR refers to this pr: https://github.com/pytorch/pytorch/pull/42174
Copy the motivation from the pr:
> Motivation: During inlinng we lose information about where the node/op is coming from in module hierarchy. The SourceDebug info is useful in providing where in the source code the definition of the node exist. One can follow the callstack to find the relevant module hierarchy. However, this is less accessible, particularly in the context of aggregating node level ops latencies to produce module level latencies. Facilitating this enables identifying which module might be expensive and can be targeted either for optimization or redesign. This also aids network architecture search approaches.

As the first step, this PR adds module instance information to inlined call stack during inlining. Next step, find the module path by traversing callstack.

### Test Plan
CI

/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=LiteInterpreterTest.DuplicatedClassTypeModuleInfo:LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo:*/LiteInterpreterTest.DuplicatedClassTypeModuleInfo/*:*/LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo --gtest_color=no


Differential Revision: [D24493220](https://our.internmc.facebook.com/intern/diff/D24493220)

[ghstack-poisoned]

cccclai committed:
Update on "[Pytorch][Annotation] Update inlined callstack with module instance info"

### Summary
This PR refers to this pr: https://github.com/pytorch/pytorch/pull/42174
Copy the motivation from the pr:
> Motivation: During inlinng we lose information about where the node/op is coming from in module hierarchy. The SourceDebug info is useful in providing where in the source code the definition of the node exist. One can follow the callstack to find the relevant module hierarchy. However, this is less accessible, particularly in the context of aggregating node level ops latencies to produce module level latencies. Facilitating this enables identifying which module might be expensive and can be targeted either for optimization or redesign. This also aids network architecture search approaches.

As the first step, this PR adds module instance information to inlined call stack during inlining. Next step, find the module path by traversing callstack.

### Test Plan
CI

/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=LiteInterpreterTest.DuplicatedClassTypeModuleInfo:LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo:*/LiteInterpreterTest.DuplicatedClassTypeModuleInfo/*:*/LiteInterpreterTest/*.DuplicatedClassTypeModuleInfo --gtest_color=no


Differential Revision: [D24493220](https://our.internmc.facebook.com/intern/diff/D24493220)

[ghstack-poisoned]
